src/ddmd/expression.d

```diff
@@ -8246,19 +8263,7 @@ extern (C++) class BinAssignExp : BinExp
             e2 = e2.castTo(sc, Type.tshiftcnt);
         }
 
-        // vectors
-        if (shift && (e1.type.toBasetype().ty == Tvector || e2.type.toBasetype().ty == Tvector))
-            return incompatibleTypes();
-
-        int isvector = type.toBasetype().ty == Tvector;
-
-        if (op == TOKmulass && isvector && !e2.type.isfloating() && (cast(TypeVector)type.toBasetype()).elementType().size(loc) != 2)
-            return incompatibleTypes(); // Only short[8] and ushort[8] work with multiply
-
-        if (op == TOKdivass && isvector && !e1.type.isfloating())
-            return incompatibleTypes();
-
-        if (op == TOKmodass && isvector)
+        if (type.toBasetype().ty == Tvector && Target.checkVectorOp(cast(TypeVector)type.toBasetype(), op, e2.type.toBasetype()))
```

Comment:
I see copy/pasta of the same code multiple times. Consider factoring it out into a separate function. In particular the constant usage of `cast` should be factored out rather than appearing everywhere, as hard cast use should be minimized.